File: android/src/main/java/com/binmatter/plugins/voskcap/VoskCapPlugin.java
package com.binmatter.plugins.voskcap;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import android.content.Context;

@CapacitorPlugin(name = "VoskCap")
public class VoskCapPlugin extends Plugin {

    private VoskCap voskCap;

    @Override
    public void load() {
        Context context = getContext();
        voskCap = new VoskCap();
        voskCap.initModel(context);
    }

    @PluginMethod
    public void startRecognition(PluginCall call) {
        call.setKeepAlive(true);

        voskCap.startListening(new VoskCap.RecognizedTextListener() {
            @Override
            public void onTextRecognized(String text) {
                JSObject ret = new JSObject();
                ret.put("text", text);
                // call.resolve(ret);
                notifyListeners("onTextRecognized", ret);
            }
        });
    }

    @PluginMethod
    public void stopRecognition(PluginCall call) {
        voskCap.stopListening();
        call.resolve();
    }
}

File: ios/Sources/VoskCapPlugin/VoskCapPlugin.swift
import Foundation
import Capacitor

@objc(VoskCapPlugin)
public class VoskCapPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "VoskCapPlugin"
    public let jsName = "VoskCap"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "startRecognition", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopRecognition", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = VoskCap()

    override public func load() {
        super.load()
        implementation.initModel()

        // Configurar onResult para notificar a los listeners de JavaScript
        implementation.onResult = { [weak self] result in
            self?.notifyListeners("onTextRecognized", data: ["text": result])
        }
    }

    @objc func startRecognition(_ call: CAPPluginCall) {
        implementation.startListening()
        call.resolve()
    }

    @objc func stopRecognition(_ call: CAPPluginCall) {
        implementation.stopListening()
        call.resolve()
    }
}
